File: src/csharp/Intel/Iced.UnitTests/Intel/InstructionTests/VATestCaseReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using Iced.Intel;

namespace Iced.UnitTests.Intel.InstructionTests {
	static class VATestCaseReader {
		public static IEnumerable<VirtualAddressTestCase> ReadFile(string filename) {
			Debug.Assert(File.Exists(filename));
			int lineNo = 0;
			foreach (var line in File.ReadLines(filename)) {
				lineNo++;
				if (line.Length == 0 || line[0] == '#')
					continue;

				VirtualAddressTestCase? tc;
				try {
					tc = ParseLine(line);
				}
				catch (Exception ex) {
					throw new Exception($"Invalid line {lineNo} ({filename}): {ex.Message}");
				}
				if (tc.HasValue)
					yield return tc.GetValueOrDefault();
			}

		}

		static readonly char[] commaSeparator = new char[] { ',' };
		static readonly char[] spaceSeparator = new char[] { ' ' };
		static readonly char[] equalSeparator = new char[] { '=' };
		static readonly char[] semicolonSeparator = new char[] { ';' };
		static VirtualAddressTestCase? ParseLine(string line) {
			var elems = line.Split(commaSeparator);
			if (elems.Length != 9)
				throw new Exception($"Invalid number of commas: {elems.Length - 1}");

			var bitness = NumberConverter.ToInt32(elems[0].Trim());
			if (CodeUtils.IsIgnored(elems[1].Trim()))
				return null;
			var hexBytes = elems[2].Trim();
			var operand = NumberConverter.ToInt32(elems[3].Trim());
			var usedMemIndex = NumberConverter.ToInt32(elems[4].Trim());
			var elementIndex = NumberConverter.ToInt32(elems[5].Trim());
			var expectedValue = NumberConverter.ToUInt64(elems[6].Trim());
			var decOptStr = elems[7].Trim();
			var decoderOptions = decOptStr == string.Empty ? DecoderOptions.None : ToEnumConverter.GetDecoderOptions(decOptStr);

			var registerValues = new List<(Register register, int elementIndex, int elementSize, ulong value)>();
			foreach (var tmp in elems[8].Split(spaceSeparator, StringSplitOptions.RemoveEmptyEntries)) {
				var kv = tmp.Split(equalSeparator);
				if (kv.Length != 2)
					throw new Exception($"Expected key=value: {tmp}");
				var key = kv[0];
				var valueStr = kv[1];

				Register register;
				int expectedElementIndex;
				int expectedElementSize;
				if (key.IndexOf(';') >= 0) {
					var parts = key.Split(semicolonSeparator);
					if (parts.Length != 3)
						throw new Exception($"Invalid number of semicolons: {parts.Length - 1}");
					register = ToEnumConverter.GetRegister(parts[0]);
					expectedElementIndex = NumberConverter.ToInt32(parts[1]);
					expectedElementSize = NumberConverter.ToInt32(parts[2]);
				}
				else {
					register = ToEnumConverter.GetRegister(key);
					expectedElementIndex = 0;
					expectedElementSize = 0;
				}
				ulong value = NumberConverter.ToUInt64(valueStr);
				registerValues.Add((register, expectedElementIndex, expectedElementSize, value));
			}

			return new VirtualAddressTestCase(bitness, hexBytes, decoderOptions, operand, usedMemIndex, elementIndex, expectedValue, registerValues.ToArray());
		}
	}
}

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/instr/VATestCaseReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.instr;

import java.util.ArrayList;

import com.github.icedland.iced.x86.CodeUtils;
import com.github.icedland.iced.x86.FileUtils;
import com.github.icedland.iced.x86.NumberConverter;
import com.github.icedland.iced.x86.ToDecoderOptions;
import com.github.icedland.iced.x86.ToRegister;
import com.github.icedland.iced.x86.dec.DecoderOptions;

final class VATestCaseReader {
	public static VirtualAddressTestCase[] readFile(String filename) {
		ArrayList<String> lines = FileUtils.readAllLines(filename);
		ArrayList<VirtualAddressTestCase> result = new ArrayList<VirtualAddressTestCase>(lines.size());
		int lineNo = 0;
		for (String line : lines) {
			lineNo++;
			if (line.length() == 0 || line.charAt(0) == '#')
				continue;

			VirtualAddressTestCase tc;
			try {
				tc = parseLine(line);
			}
			catch (Exception ex) {
				throw new UnsupportedOperationException(String.format("Invalid line %d (%s): %s", lineNo, filename, ex.getMessage()));
			}
			if (tc != null)
				result.add(tc);
		}
		return result.toArray(new VirtualAddressTestCase[result.size()]);
	}

	private static VirtualAddressTestCase parseLine(String line) {
		String[] elems = line.split(",");
		if (elems.length != 9)
			throw new UnsupportedOperationException(String.format("Invalid number of commas: %d", elems.length - 1));

		int bitness = NumberConverter.toInt32(elems[0].trim());
		if (CodeUtils.isIgnored(elems[1].trim()))
			return null;
		String hexBytes = elems[2].trim();
		int operand = NumberConverter.toInt32(elems[3].trim());
		int usedMemIndex = NumberConverter.toInt32(elems[4].trim());
		int elementIndex = NumberConverter.toInt32(elems[5].trim());
		long expectedValue = NumberConverter.toUInt64(elems[6].trim());
		String decOptStr = elems[7].trim();
		int decoderOptions = decOptStr.equals("") ? DecoderOptions.NONE : ToDecoderOptions.get(decOptStr);

		ArrayList<VARegisterValue> registerValues = new ArrayList<VARegisterValue>();
		for (String tmp : elems[8].split(" ")) {
			if (tmp.length() == 0)
				continue;
			String[] kv = tmp.split("=");
			if (kv.length != 2)
				throw new UnsupportedOperationException(String.format("Expected key=value: %s", tmp));
			String key = kv[0];
			String valueStr = kv[1];

			int register;
			int expectedElementIndex;
			int expectedElementSize;
			if (key.indexOf(';') >= 0) {
				String[] parts = key.split(";");
				if (parts.length != 3)
					throw new UnsupportedOperationException(String.format("Invalid number of semicolons: %d", parts.length - 1));
				register = ToRegister.get(parts[0]);
				expectedElementIndex = NumberConverter.toInt32(parts[1]);
				expectedElementSize = NumberConverter.toInt32(parts[2]);
			} else {
				register = ToRegister.get(key);
				expectedElementIndex = 0;
				expectedElementSize = 0;
			}
			long value = NumberConverter.toUInt64(valueStr);
			registerValues.add(new VARegisterValue(register, expectedElementIndex, expectedElementSize, value));
		}

		return new VirtualAddressTestCase(bitness, hexBytes, decoderOptions, operand, usedMemIndex, elementIndex, expectedValue,
				registerValues.toArray(new VARegisterValue[registerValues.size()]));
	}
}
